leetcode_bwang

```diff
@@ -0,0 +1,28 @@
+class Solution {
+    public List<List<Integer>> permuteUnique(int[] nums) {
+        List<List<Integer>> wraplist = new ArrayList<>();
+        List<Integer> sublist = new ArrayList<>();
+        boolean[] visited = new boolean[nums.length];
+        Arrays.sort(nums);
+        dfs(wraplist, sublist, nums, visited);
+        return wraplist;
+    }
+    
+    public void dfs(List<List<Integer>> wraplist, List<Integer> sublist, int[] nums, boolean[] visited) {
+        if (sublist.size() == nums.length) {
+            wraplist.add(new ArrayList<>(sublist));
+            return;
+        }
+        
+        for (int i = 0; i < nums.length; i++) {  
+            if (i > 0 && nums[i] == nums[i - 1] && !visited[i - 1]) continue;
+            if (!visited[i]) {
+                sublist.add(nums[i]);
+                visited[i] = true;
+                dfs(wraplist, sublist, nums, visited);
+                sublist.remove(sublist.size() - 1);
+                visited[i] = false;
+            }
+        }
+    }
```

Comment:
Here's my solution, you can take a look as a reference. Let me write some general ideas for this kind of problems.

Generally, the requirement wants you to return all of possible permuted numbers placed in lists. You can solve it step by step, first don't consider duplicate cases, output all of permutations with duplicates. 

Usually, we use depth first search to solve it in a recursive way.  For recursive calls, you wanna consider the base case first. Under which condition do you want this recursive call to end and return? In our case, it's when you put all of numbers in a list container, so the size of list needs to equals total count of numbers. And then you wanna add number elements into list one by one. Since we are going to add it recursively, we'll need to have some way to keep track of which number has been added in previous calls. There are two ways of doing it: use an index and increment that index in subsequent calls or a boolean array to cache if the element is visited or not. At each call, we wanna loop over each element within the array, if it's not been visited before, add it to your list container, mark it as visited. When backtracking, reset your previous actions. 

Next, we'll work on duplicates removal.